### PR TITLE
Ufunc support

### PR DESCRIPTION
**Pull request by alexnick83**
**NOTE**: Actual ufuncs will be added in subsequent PR's

Support for ufunc functionality:
- [x] brodcasting
- [ ] ~~output type determination~~ (use the current python-op implementation for now)
- [ ] ~~casting rules~~ (use the current python-op implementation for now)

Support for ufunc keyword arguments (potential support for keywords applying to generalized ufuncs, such as axes, axis, keepdims, will be evaluated later):
- [x] out
- [x] where
- [x] dtype

Support for ufunc methods:
- [x] reduce (without 'where' and 'dtype' support - the former has to be done in the reduce library node, while the latter doesn't make sense)
- [x] accumulate
- [ ] ~~reduceat~~ (too complicated for a very niche use, re-evaluate a potential implementation later)
- [x] outer
- [ ] ~~at~~ (also re-evaluate potential implementation later)

### CODE REVIEW

File: dace/frontend/python/replacements.py [hunk 1]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+            tasklet_params['name'],
+            map_indices,
+            input_memlets,
+            tasklet_params['code'],
+            output_memlets,
+            external_edges=True
+        )
+
+
+@oprepo.replaces_ufunc('ufunc')
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2247]: Not necessary (is this an actual ufunc name?)

File: dace/frontend/python/replacements.py [hunk 2]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+                if isinstance(inpdata, data.Array):
+                    state.add_mapped_tasklet(
+                        name=state.label + "_reduce_initial",
+                        map_ranges={
+                            "__i{i}".format(i=i): "0:{s}".format(s=s)
+                            for i, s in enumerate(inpdata.shape)
+                            if i not in axis
+                        },
+                        inputs={
+                            "__inp": dace.Memlet.simple(inputs[0], ','.join([
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2558]: could you please use the new dace.Memlet() constructor? it should be easier to work with.

File: dace/frontend/python/replacements.py [hunk 3]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+                                "0" if i in axis else "__i{i}".format(i=i)
+                                for i in range(len(inpdata.shape))
+                            ]))
+                        },
+                        outputs={
+                            "__out": dace.Memlet.simple(
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2564]: same

File: dace/frontend/python/replacements.py [hunk 4]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+                            ]))
+                        },
+                        code="__out = __inp",
+                        external_edges=True
+                    )
+                else:
+                    r = state.add_read(inputs[0])
+                    w = state.add_write(intermediate_name)
+                    state.add.nedge(
+                        r, w, dace.Memlet.simple.from_array(inpdata))
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2578]:
```suggestion
                        r, w, dace.Memlet.from_array(inpdata))
```

File: dace/frontend/python/replacements.py [hunk 5]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+    # Create subgraph
+    if isinstance(inputs[0], str) and inputs[0] in sdfg.arrays.keys():
+        _reduce(sdfg, state, ufunc_impl['reduce'], inputs[0], intermediate_name,
+                axis=axis, identity=initial)
+    else:
+        tasklet = state.add_tasklet(state.label + "_tasklet", {}, {'__out'},
+                                    "__out = {}".format(inputs[0]))
+        out_node = state.add_write(intermediate_name)
+        datadesc = sdfg.arrays[intermediate_name]
+        state.add_edge(tasklet, '__out', out_node, None, dace.Memlet.simple(
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2592]: same

File: dace/frontend/python/replacements.py [hunk 6]
```diff
@@ -1582,3 +1592,1287 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+    if keepdims:
+        intermediate_node = None
+        for n in state.nodes():
+            if isinstance(n, nodes.AccessNode) and n.data == intermediate_name:
+                intermediate_node = n
+                break
+        if not intermediate_node:
+            raise Exception("Something went wrong!")
+        out_node = state.add_write(outputs[0])
+        state.add_nedge(intermediate_node, out_node, dace.Memlet.simple(
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2604]: same

File: dace/frontend/python/replacements.py [hunk 7]
```diff
@@ -1582,3 +1592,1292 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+            intermediate_name, subsets.Range.from_array(datadesc)))
+    
+    if keepdims:
+        intermediate_node = None
+        for n in state.nodes():
+            if isinstance(n, nodes.AccessNode) and n.data == intermediate_name:
+                intermediate_node = n
+                break
+        if not intermediate_node:
+            raise Exception("Something went wrong!")
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2602]: raise ValueError please

File: dace/frontend/python/replacements.py [hunk 8]
```diff
@@ -1582,3 +1592,1292 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+    outconn = nested_sdfg.temp_data_name()
+    shape = [datadesc.shape[axis]]
+    strides = [datadesc.strides[axis]]
+    nested_sdfg.add_array(inpconn, shape, result_type, strides=strides)
+    nested_sdfg.add_array(outconn, shape, result_type, strides=strides)
+    
+    init_state = nested_sdfg.add_state(label="init")
+    r = init_state.add_read(inpconn)
+    w = init_state.add_write(outconn)
+    init_state.add_nedge(r, w, dace.Memlet.simple(
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2708]: same

File: dace/frontend/python/replacements.py [hunk 9]
```diff
@@ -1582,3 +1592,1292 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+    )
+
+    loop_idx = "__i{}".format(axis)
+    loop_idx_m1 = "__i{} - 1".format(axis)
+    body_state.add_edge(r1, None, t, '__in1',
+                        dace.Memlet.simple(inpconn, loop_idx))
+    body_state.add_edge(r2, None, t, '__in2',
+                        dace.Memlet.simple(outconn, loop_idx_m1))
+    body_state.add_edge(t, '__out', w, None,
+                        dace.Memlet.simple(outconn, loop_idx))
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2729]: same

File: dace/frontend/python/replacements.py [hunk 10]
```diff
@@ -1582,3 +1592,1292 @@ def _matmult(visitor, sdfg: SDFG, state: SDFGState, op1: str, op2: str):
...
+    w = state.add_write(outputs[0])
+    codenode = state.add_nested_sdfg(nested_sdfg, sdfg,
+                                        {inpconn}, {outconn})
+    me, mx = state.add_map(state.label + '_map', map_range)
+    state.add_memlet_path(r, me, codenode,
+                          memlet=dace.Memlet.simple(inputs[0], input_idx),
+                          dst_conn=inpconn)
+    state.add_memlet_path(codenode, mx, w,
+                          memlet=dace.Memlet.simple(outputs[0], output_idx),
+                          src_conn=outconn)
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2749]: same

File: tests/numpy/ufunc_support_test.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
+import dace
```
Comment (tbennun) [inline tests/numpy/ufunc_support_test.py:1]: Copyright line

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/frontend/python/replacements.py (hunk 1 above)
Comment (alexnick83) [reply dace/frontend/python/replacements.py:2247]: No, all ufuncs will use this single method. I used a replacement operator, so I can have different functions for each one of the ufunc methods (currently only reduce, accumulate, and outer supported).

**Review by tbennun — APPROVED** [review]

**Pull request merged**